src/vec.rs

```diff
@@ -1175,6 +1241,39 @@ where
     }
 }
 
+#[macro_export]
+/// Shorthand for defining vectors using array syntax.
+///
+/// Create `Vec` with a list of elements and capacity:
+/// ```
+/// let v = heapless::vec![1, 2, 3; 7];
+/// assert_eq!(v.as_slice(), &[1, 2, 3]);
+/// assert_eq!(v.capacity(), 7);
+/// ```
+///
+/// Create `Vec` with a repeated element, length, and capacity:
+/// ```
+/// let v = heapless::vec!['a'; 3; 4];
+/// assert_eq!(v.as_slice(), &['a', 'a', 'a']);
+/// assert_eq!(v.capacity(), 4);
+/// ```
+///
+/// Unlike the `std` version of this macro, the repeat element must be `Copy` or a constant, like
+/// with repeat elements in array expressions.
+macro_rules! vec {
+    (; $cap:expr) => {
+        heapless::Vec::<_, $cap>::new()
+    };
+
+    ($($elem:expr),+ ; $cap:expr) => {
```

Comment:
this should be consistent with std. 

As it's written now, `vec![42; 5]` in std it would create a vec with five 42's, while in heapless it creates a vec with a single 42 with a capacity of 5.

IMO the macro should behave like this:

- `vec![1,2,3]`: 3 elements, capacity is inferred.
- `vec![42; 3]`: element repeated 3 times, capacity is inferred.

and then invent some extra syntax for specifying capacity explicitly that doesn't conflict with that. 

In #369 we had the same issue, we went with `format!(256; "foo {}", 42)` where `256` is the capacity, but that doesn't work well here...